core: Add new function libusb_get_session_data

vovkos opened the pull request:
This trivial getter returns the backend-specific handle (libusb_device::session_data)

Allows finding the original node for a particular libusb_device in the system device tree (DEVINST in SetupAPI/CfgMgr32 on Windows; sessionID in IOKit on macOS) and accessing it if necessary.

mcuee commented:
Please help to provide a use case of this API, thanks.

vovkos commented:
On Windows & macOS, this allows getting device descriptors (manufacturer, product name, serial number) without actually opening the corresponding `libusb_device`. Opening a device *just to get information* is not really the best approach anyway -- on Windows & macOS, you can't simultaneously open a USB device from two processes, so opening a device to read its descriptors could interfere with another process that also wants to work with the same device. Moreover, on Windows, you can only open a `libusb`-compatible driver (such as `WinUSB`) -- so you can't even get string descriptors for all other devices. Moreover, once you find a device in the system tree using this `libusb_get_session_data`, it becomes possible to get many other pieces of system information (e.g., driver path, registry, hardware IDs, device instance ID, etc).

A comparison of USB device enumeration on Windows done with `libusb` alone vs `libusb` + `SetupAPI`:

```
ENUMERATING WITH LIBUSB
=======================
device #0
  bus:            3
  address:        1
  port:           2
  VID:            0x046d
  PID:            0xc31c
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Low-Speed (1.5 Mbps)
  cannot open device (No such device (it may have been disconnected))

device #1
  bus:            3
  address:        2
  port:           3
  VID:            0x14a5
  PID:            0xa011
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Full-Speed (12 Mbps)
  manufacturer*:  ?
  product*:       
  serial number*: NONE

device #2
  bus:            6
  address:        1
  port:           5
  VID:            0x0529
  PID:            0x0620
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Full-Speed (12 Mbps)
  cannot open device (Operation not supported or unimplemented on this platform)

device #3
  bus:            1
  address:        2
  port:           4
  VID:            0x09fb
  PID:            0x6001
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Full-Speed (12 Mbps)
  cannot open device (Operation not supported or unimplemented on this platform)

device #4
  bus:            7
  address:        1
  port:           4
  VID:            0x046d
  PID:            0x092e
  class:          0xff - Vendor-specific
  subclass:       0xff
  speed:          Full-Speed (12 Mbps)
  cannot open device (Operation not supported or unimplemented on this platform)

ENUMERATING WITH LIBUSB + SETUPAPI
===========================
device #0
  bus:            3
  address:        1
  port:           2
  VID:            0x046d
  PID:            0xc31c
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Low-Speed (1.5 Mbps)
  description:    USB Composite Device
  manufacturer:   (Standard USB Host Controller)
  driver:         \SystemRoot\System32\drivers\usbccgp.sys
  manufacturer*:  Logitech
  product*:       USB Keyboard

device #1
  bus:            3
  address:        2
  port:           3
  VID:            0x14a5
  PID:            0xa011
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Full-Speed (12 Mbps)
  description:    USB Input Device
  manufacturer:   (Standard system devices)
  driver:         \SystemRoot\System32\drivers\hidusb.sys
  manufacturer*:  USB Gaming Mouse
  product*:       USB Gaming Mouse

device #2
  bus:            6
  address:        1
  port:           5
  VID:            0x0529
  PID:            0x0620
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Full-Speed (12 Mbps)
  description:    USB Token
  manufacturer:   Aladdin Knowledge Systems Ltd.
  driver:         \SystemRoot\system32\drivers\aksup.sys
  manufacturer*:  SafeNet
  product*:       Token JC

device #3
  bus:            1
  address:        2
  port:           4
  VID:            0x09fb
  PID:            0x6001
  class:          0x00 - Multi-function
  subclass:       0x00
  speed:          Full-Speed (12 Mbps)
  description:    Altera USB-Blaster
  manufacturer:   Altera
  driver:         \SystemRoot\system32\drivers\usbblstr.sys
  manufacturer*:  Altera
  product*:       USB-Blaster
  serial number*: 91d28408

device #4
  bus:            7
  address:        1
  port:           4
  VID:            0x046d
  PID:            0x092e
  class:          0xff - Vendor-specific
  subclass:       0xff
  speed:          Full-Speed (12 Mbps)
  description:    Logitech QuickCam Chat
  manufacturer:   Logitech
  driver:         \SystemRoot\system32\DRIVERS\LV561V64.SYS
  manufacturer*:          
  product*:       Camera
```

Also, it's worth mentioning that the descriptors on a device could provide less information than the properties of a device tree node. For example, see the last entry. The device itself gives a single string descriptor saying "Camera", while the Windows driver `.inf` files provide much better information -- "Logitech QuickCam Chat".

A screenshot showing the application of this new function `libusb_get_session_data` to provide detailed information on each USB device in our `libusb`-based [USB Endpoint Terminal](https://ioninja.com/plugins/usb-endpoint.html):

![image](https://github.com/libusb/libusb/assets/1272029/675c83af-af93-4db7-9be9-2e70e76353e3)

Youw commented:
Sounds like a useful workaround, but long-term I'd make my bets on #1258

vovkos commented:
As I said, it's not *only* about retrieving string descriptors without opening a device. Being able to find the original node in the system device tree is generally helpful. OK, another realistic example of when it could be necessary is to traverse up the device tree and find the corresponding USB root hub this device is attached to (so that you can start a USBPcap session on this particular root hub).

Just like `QWidget` provides [`winId()`](https://doc.qt.io/qt-6/qwidget.html#winId) or `QFile` provides [`handle()`](https://doc.qt.io/qt-6/qfiledevice.html#handle), `libusb_device` should provide means of finding the original system device.

mcuee commented:
@tormodvolden 

This PR seems to be useful for some use cases. Please review as well. Thanks.

Youw commented:
> Just like QWidget provides [winId()](https://doc.qt.io/qt-6/qwidget.html#winId) or QFile provides [handle()](https://doc.qt.io/qt-6/qfiledevice.html#handle), libusb_device should provide means of finding the original system device.

The danger if having such APIs is that it makes it easy to break the library's invariant.
For (almost) each use-case it is possible to provide an alternative fine-grained API (like walkng the parents, etc.).
Worst-case scenario - prividing an _identifier_ of the underlying OS device (e.g. a device path on Windows or LocationID on macOS) is a safer choice than providing the indernal handle itself.

vovkos commented:
> The danger if having such APIs is that it makes it easy to break the library's invariant. For (almost) each use-case it is possible to provide an alternative fine-grained API (like walkng the parents, etc.). Worst-case scenario - prividing an _identifier_ of the underlying OS device (e.g. a device path on Windows or LocationID on macOS) is a safer choice than providing the indernal handle itself

Hmm, I can't really see how exposing a location in the system device tree could possibly interfere with the operation of `libusb`. Unless I'm missing something, of course -- in which case, please provide an example scenario. 

What we are going to expose is *not an open file/device handle* -- instead, it's a *location identifier*. In this regard, it's even safer than the aforementioned `QFile::handle()` or `QWidget::winId()` (that provide an actual handle). Also, exposing a string path (such as an `InstanceID`/`InterfacePath`/`locationID`) is not going to be any different in terms of what can be done with it (i.e., you can unambiguously restore paths from `DEVINST` or `sessionID` and vice versa).

As for creating multiple sub-APIs for different tasks -- sure, but it's a much more challenging and time/effort consuming endeavor compared to a single trivial getter that solves all that *right now* (yes, including getting descriptors without opening a device).

Youw reviewed (commented):
> What we are going to expose is not an open file/device handle -- instead, it's a location identifier.

That does makes it a lesser evil, I was confused by the documentation.

vovkos commented on libusb/core.c:
```diff
@@ -902,6 +903,17 @@ void API_EXPORTED libusb_free_device_list(libusb_device **list,
+ * Get the backend-specific handle (session data in terms of libusb) of a device.
+ * Can be used for direct low-level access to the backend when necessary.
+ * \param dev a device
+ * \returns the backend-specific handle (session data)
```
> lets avoid the term `handle` - it might be confusing

Good point. It's not a handle, indeed. I tried to rewrite the comment and make it more clear. I'm not a native English speaker, though. It could make sense for a native speaker to review/edit this comment (it goes to docs if merged).

Youw replied:
> I'm not a native English speaker

Neither do I, but the updated version is much better, thanks.

Youw reviewed (approved):
This looks like a cheap workaround for some potential, otherwise much more complicated, problems.
Leaving it to USB Core team for final approve/merge.

Youw commented on libusb/core.c:
```diff
@@ -902,6 +903,24 @@ void API_EXPORTED libusb_free_device_list(libusb_device **list,
 	free(list);
 }
 
+/** \ingroup libusb_dev
```
oh, and the documentation misses `since version` clause
check for instance `libusb_get_port_numbers`

vovkos replied:
Sure. I used the next upcoming version and build number (v1.0.27, `LIBUSB_API_VERSION >= 0x0100010B`), correct?

Also, removed the  `LIBUSB_HAS_GET_SESSION_DATA` preprocessor constant (users should check `LIBUSB_API_VERSION` for conditional compilation, just like with the rest of libusb API).

vovkos commented:
This last force-push is just to sign the commits (apparently, this is required?); there are no actual changes.

mcuee commented:
> This last force-push is just to sign the commits (apparently, this is required?); there are no actual changes.

Not required for libusb project.

mcuee commented:
@tormodvolden and @sonatique 

Just wonder if we stiill need this if #1617 gets merged.

* #1617

sonatique commented:
@mcuee  that is a good question. I need to look into more details into the code. I am wondering why #1617 did not mention this.

I guess the one who can really answer the question is @vovkos . @vovkos  does #1617 fulfil the need you tried to cover with this PR?

vovkos commented:
@sonatique, @mcuee,
 
I'd put it another way -- do we even need #1617 if this gets merged? ;)

The current PR (2 years old already ;) provides information that is equivalent to what #1617 provides. But:

1) It's simpler: one trivial integer getter added to the API, no backend changes vs two new API string calls + modification of backend method table + implementation for all backends, 12 total files changed

2) It's more efficient: returning an integer vs allocating/deallocating strings; clients of the new API will be matching the device by integer ID vs searching the device by string path

3) It's tested in prod: IO Ninja (USB Endpoint/USB Monitor/HID Terminal/HID Monitor) is using this patch to match and extract all the USB device information from the system device tree on Windows and macOS (on Linux, the problem of matching libusb device to sysfs doesn't exist in the first place).

Mr-Bossman commented:
> @vovkos I think you definitively have a point here.
> 
> I have to say that I didn't follow things 2 years ago and didn't try to trace back how we ended in this situation with 2 concurrent PRs.
> 
> Yours seems indeed simpler and indeed has all the benefits you listed.
> 
> To be honest I don't see which limitations of your PR #1617 did try to overcome.
> 
> I would tend to favor simplicity and merge your PR, but I probably don't understand the big picture.
> 
> Maybe there are scenarios that only #1617 can address and we'll need to have both PR merged if the overlap is not too big.
> 
> @Mr-Bossman : what is your opinion here? Thank you in advance.

I'll take a look I didn't know about this PR before. I have been a bit busy with school so I may be slow.

vovkos commented on libusb/core.c:
```diff
@@ -902,6 +903,17 @@ void API_EXPORTED libusb_free_device_list(libusb_device **list,
+/** \ingroup libusb_dev
+ * Get the backend-specific handle (session data in terms of libusb) of a device.
+ * Can be used for direct low-level access to the backend when necessary.
+ * \param dev a device
```
Re `dev` not being null -- that's a valid point; can add this. But then it will become inconsistent with docs for the rest of the libusb API. Pretty much all libusb functions that take `libusb_device*` assume it's non-NULL and simply dereference it without checking (or mentioning it in the docs). I think I used `libusb_get_bus_number` (or something like that) as a template for both code & docs.

Re squashing & rebasing -- absolutely! When merging, all the staged changes can (and should) be rebased and squashed into a single commit. I believe it wouldn't cause any conflicts due to the trivial nature of this PR.

vovkos committed:
core: Add new function libusb_get_session_data

This trivial getter returns the backend-specific handle (libusb_device::session_data)

Allows finding the original node for a particular libusb_device in the system device tree (DEVINST in SetupAPI/CfgMgr32 on Windows; sessionID in IOKit on macOS) and accessing it if necessary.

sonatique commented:
@mcuee @Youw @seanm @tormodvolden : I see no reason no to merge this PR now. Any objection?

Youw reviewed (approved):
LGTM

Mr-Bossman commented:
> @tormodvolden and @sonatique
> 
> Just wonder if we stiill need this if #1617 gets merged.
> 
> * [core: Return an internal handle for deeper integration #1617](https://github.com/libusb/libusb/pull/1617)

No probably not.

Though I will say that the format for `session_data` is dependent on the back-ends use of it not changing in the future, but i don't see that happening.

mcuee commented:
Testing using MSYS2 mingw64 under Windows 11.

```
 MINGW64 /c/work/libusb/libusb_pr1314
$ make check
Making check in libusb
make[1]: Entering directory '/c/work/libusb/libusb_pr1314/libusb'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/c/work/libusb/libusb_pr1314/libusb'
Making check in examples
make[1]: Entering directory '/c/work/libusb/libusb_pr1314/examples'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/c/work/libusb/libusb_pr1314/examples'
Making check in tests
make[1]: Entering directory '/c/work/libusb/libusb_pr1314/tests'
make  check-TESTS
make[2]: Entering directory '/c/work/libusb/libusb_pr1314/tests'
make[3]: Entering directory '/c/work/libusb/libusb_pr1314/tests'
PASS: stress.exe
PASS: stress_mt.exe
PASS: set_option.exe
PASS: init_context.exe
============================================================================
Testsuite summary for libusb-1.0 1.0.29
============================================================================
# TOTAL: 4
# PASS:  4
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[3]: Leaving directory '/c/work/libusb/libusb_pr1314/tests'
make[2]: Leaving directory '/c/work/libusb/libusb_pr1314/tests'
make[1]: Leaving directory '/c/work/libusb/libusb_pr1314/tests'
make[1]: Entering directory '/c/work/libusb/libusb_pr1314'
make[1]: Leaving directory '/c/work/libusb/libusb_pr1314'
```

mcuee commented:
Testing under Ubuntu 24.04 Linux.
```
mcuee@PCRyzen5600G:~/build/libusb/libusb_pr1314$ make check
Making check in libusb
make[1]: Entering directory '/home/mcuee/build/libusb/libusb_pr1314/libusb'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/mcuee/build/libusb/libusb_pr1314/libusb'
Making check in examples
make[1]: Entering directory '/home/mcuee/build/libusb/libusb_pr1314/examples'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/mcuee/build/libusb/libusb_pr1314/examples'
Making check in tests
make[1]: Entering directory '/home/mcuee/build/libusb/libusb_pr1314/tests'
make  check-TESTS
make[2]: Entering directory '/home/mcuee/build/libusb/libusb_pr1314/tests'
make[3]: Entering directory '/home/mcuee/build/libusb/libusb_pr1314/tests'
PASS: stress
PASS: stress_mt
PASS: set_option
PASS: init_context
PASS: umockdev
============================================================================
Testsuite summary for libusb-1.0 1.0.29
============================================================================
# TOTAL: 5
# PASS:  5
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[3]: Leaving directory '/home/mcuee/build/libusb/libusb_pr1314/tests'
make[2]: Leaving directory '/home/mcuee/build/libusb/libusb_pr1314/tests'
make[1]: Leaving directory '/home/mcuee/build/libusb/libusb_pr1314/tests'
make[1]: Entering directory '/home/mcuee/build/libusb/libusb_pr1314'
make[1]: Leaving directory '/home/mcuee/build/libusb/libusb_pr1314'

```

mcuee reviewed (approved):
Approved based on testing results.

mcuee commented:
@sonatique 

I have approved this PR.
